File: node/src/sentiment.ts
/**
 * sentiment.ts — sentiment tools. Mirrors ../../python/ai_trader_tools/sentiment.py.
 *
 * getFearGreedIndex: REAL, no API key — free alternative.me endpoint.
 * fetchNewsHeadlines: pluggable provider, bring your own API key.
 *
 * NOTE: this file does not include a VADER-equivalent text-sentiment function out of the
 * box — Node's ecosystem doesn't have a drop-in VADER port as battle-tested as Python's.
 * Recommended: install the `vader-sentiment` npm package yourself, or call the Python
 * analyze_text_sentiment tool via the REST API / MCP server from your Node app instead.
 */
import fetch from "node-fetch";

export async function getFearGreedIndex(limit = 1) {
  const resp = await fetch(`https://api.alternative.me/fng/?limit=${limit}`);
  if (!resp.ok) throw new Error(`alternative.me request failed: ${resp.status}`);
  const json: any = await resp.json();
  return {
    source: "alternative.me",
    readings: (json.data || []).map((d: any) => ({
      value: parseInt(d.value, 10),
      classification: d.value_classification,
      timestamp: d.timestamp,
    })),
  };
}

async function fetchNewsApi(query: string, apiKey: string, limit: number) {
  const url = `https://newsapi.org/v2/everything?q=${encodeURIComponent(query)}&sortBy=publishedAt&pageSize=${limit}&apiKey=${apiKey}`;
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`newsapi request failed: ${resp.status}`);
  const json: any = await resp.json();
  return (json.articles || []).map((a: any) => ({ title: a.title, source: a.source?.name, publishedAt: a.publishedAt, url: a.url }));
}

async function fetchCryptoPanic(query: string, apiKey: string, limit: number) {
  const url = `https://cryptopanic.com/api/v1/posts/?auth_token=${apiKey}&currencies=${encodeURIComponent(query)}&public=true`;
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`cryptopanic request failed: ${resp.status}`);
  const json: any = await resp.json();
  return (json.results || []).slice(0, limit).map((p: any) => ({ title: p.title, source: p.source?.title, publishedAt: p.published_at, url: p.url }));
}

const PROVIDERS: Record<string, (q: string, k: string, l: number) => Promise<any[]>> = {
  newsapi: fetchNewsApi,
  cryptopanic: fetchCryptoPanic,
};

export async function fetchNewsHeadlines(query: string, apiKey: string, provider = "newsapi", limit = 20) {
  const fn = PROVIDERS[provider];
  if (!fn) throw new Error(`Unknown provider '${provider}'. Available: ${Object.keys(PROVIDERS).join(", ")}`);
  const headlines = await fn(query, apiKey, limit);
  return { provider, query, headlines };
}

File: python/ai_trader_tools/sentiment.py
"""
sentiment.py — sentiment analysis tools.

- get_fear_greed_index: REAL, no API key — hits the free alternative.me Fear & Greed API.
- analyze_text_sentiment: REAL, no API key — VADER lexicon sentiment (pip install vaderSentiment).
- fetch_news_headlines: pluggable provider interface. Bring your own API key for NewsAPI or
  CryptoPanic (or add your own provider — see PROVIDERS below).
"""
from typing import List, Optional

import requests

try:
    from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
    _analyzer = SentimentIntensityAnalyzer()
except ImportError:
    _analyzer = None


def get_fear_greed_index(limit: int = 1) -> dict:
    resp = requests.get("https://api.alternative.me/fng/", params={"limit": limit}, timeout=10)
    resp.raise_for_status()
    data = resp.json().get("data", [])
    return {
        "source": "alternative.me",
        "readings": [
            {"value": int(d["value"]), "classification": d["value_classification"], "timestamp": d["timestamp"]}
            for d in data
        ],
    }


def analyze_text_sentiment(texts: List[str]) -> dict:
    if _analyzer is None:
        raise ImportError("vaderSentiment is required — install with: pip install vaderSentiment")
    results = []
    total = 0.0
    for text in texts:
        scores = _analyzer.polarity_scores(text)
        results.append({"text": text, "compound": scores["compound"], "pos": scores["pos"], "neu": scores["neu"], "neg": scores["neg"]})
        total += scores["compound"]
    avg = total / len(texts) if texts else 0.0
    label = "bullish" if avg > 0.15 else "bearish" if avg < -0.15 else "neutral"
    return {"results": results, "average_compound": round(avg, 4), "label": label}


def _fetch_newsapi(query: str, api_key: str, limit: int) -> list:
    resp = requests.get(
        "https://newsapi.org/v2/everything",
        params={"q": query, "sortBy": "publishedAt", "pageSize": limit, "apiKey": api_key},
        timeout=10,
    )
    resp.raise_for_status()
    articles = resp.json().get("articles", [])
    return [{"title": a["title"], "source": a["source"]["name"], "publishedAt": a["publishedAt"], "url": a["url"]} for a in articles]


def _fetch_cryptopanic(query: str, api_key: str, limit: int) -> list:
    resp = requests.get(
        "https://cryptopanic.com/api/v1/posts/",
        params={"auth_token": api_key, "currencies": query, "public": "true"},
        timeout=10,
    )
    resp.raise_for_status()
    posts = resp.json().get("results", [])[:limit]
    return [{"title": p["title"], "source": p.get("source", {}).get("title"), "publishedAt": p["published_at"], "url": p["url"]} for p in posts]


PROVIDERS = {"newsapi": _fetch_newsapi, "cryptopanic": _fetch_cryptopanic}


def fetch_news_headlines(query: str, api_key: str, provider: str = "newsapi", limit: int = 20) -> dict:
    if provider not in PROVIDERS:
        raise ValueError(f"Unknown provider '{provider}'. Available: {list(PROVIDERS)}. Add your own in PROVIDERS.")
    headlines = PROVIDERS[provider](query, api_key, limit)
    return {"provider": provider, "query": query, "headlines": headlines}
